src/plugins/gs-plugin-eos.c

```diff
@@ -264,6 +265,155 @@ app_is_renamed (GsApp *app)
+static gchar *
+locale_to_language (const gchar *locale)
+{
+	gchar *tmp;
+	gchar *country_code;
+
+	/* invalid */
+	if (locale == NULL)
+		return NULL;
+
+	/* return the part before the _ (not always 2 chars!) */
+	country_code = g_strdup (locale);
+	tmp = g_strstr_len (country_code, -1, "_");
+	if (tmp != NULL)
+		*tmp = '\0';
+	return country_code;
+}
+
+static gboolean
+gs_plugin_locale_is_compatible (GsPlugin *plugin,
+				const char *locale)
+{
+	g_autofree char *lang = NULL;
+	const char *plugin_locale = gs_plugin_get_locale (plugin);
+	const char *plugin_lang = gs_plugin_get_language (plugin);
+
+	/* if we don't know the current locale than all locales should be
+	 * compatible */
+	if (!plugin_locale)
+		return TRUE;
+
+	if (g_strcmp0 (locale, plugin_locale) == 0)
+		return TRUE;
+
+	lang = locale_to_language (locale);
+
+	/* in case one of the given locales only represents the language
+	 * (e.g. 'es') check if it's compatible with the other locale's language
+	 * component */
+	if ((g_strcmp0 (lang, plugin_locale) == 0) ||
+	    (g_strcmp0 (plugin_lang, locale) == 0))
+		return TRUE;
+
+	return FALSE;
+}
+
+static char *
+get_app_locale_cache_name (const char *name)
+{
+	g_autofree char *locale_cache_name = g_strdup (name);
+	guint name_length = strlen (locale_cache_name);
+	char *locale_split = NULL;
+
+	/* locales can be as long as 5 chars (e.g. pt_PT) so  */
+	if (name_length <= 5)
+		return NULL;
+
+	locale_split = g_strrstr (locale_cache_name + name_length - 5, "_");
+	if (locale_split)
+		*locale_split = '\0';
+
+	return g_strdup_printf ("locale:%s", locale_cache_name);
+}
+
+static gboolean
+gs_plugin_locale_cached_app_is_best_match (GsPlugin *plugin,
+					   const char *locale_app_id)
+{
+	GsApp *cached_app = NULL;
+	cached_app = gs_plugin_cache_lookup (plugin, locale_app_id);
+
+	if (!cached_app)
+		return FALSE;
+
+	return g_str_has_suffix (gs_app_get_flatpak_name (cached_app),
+				 gs_plugin_get_locale (plugin));
+}
+
+static void
+gs_plugin_update_locale_cache_app (GsPlugin *plugin,
+				   const char *locale_app_id,
+				   GsApp *app)
+{
+	GsApp *cached_app = gs_plugin_cache_lookup (plugin, locale_app_id);
+
+	if (cached_app) {
+		/* avoid blacklisting the same app that's already cached */
+		if (cached_app == app)
+			return;
+
+		gs_app_add_category (cached_app, "Blacklisted");
+	}
+
+	gs_plugin_cache_add (plugin, locale_app_id, app);
+}
+
+static gboolean
+gs_plugin_eos_blacklist_kapp_if_needed (GsPlugin *plugin, GsApp *app)
+{
+	const guint endless_prefix_len = strlen (ENDLESS_ID_PREFIX);
+	g_autofree char *locale_cache_name = NULL;
+	g_auto(GStrv) tokens = NULL;
+	const char *last_token = NULL;
+	const char *locale = gs_plugin_get_locale (plugin);
+	guint num_tokens = 0;
+	/* getting the app name, besides skipping the '.desktop' part of the id
+	 * also makes sure we're dealing with a Flatpak app */
+	const char *app_name = gs_app_get_flatpak_name (app);
+
+	if (!app_name || !locale ||
```

Comment:
I don't think `locale` can ever be `NULL` here. Other callers also assume it's not NULL